drv/fpga-api/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! API crate for the FPGA server.
+
+#![no_std]
+
+use drv_spi_api::SpiError;
+use userlib::*;
+use zerocopy::AsBytes;
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+#[repr(u16)]
+pub enum FpgaError {
+    ImplError(u8),
+    BitstreamError(u8),
+    InvalidState,
+    InvalidValue,
+    PortDisabled,
+}
+
+impl From<FpgaError> for u16 {
+    fn from(e: FpgaError) -> Self {
+        match e {
+            FpgaError::ImplError(error_code) => 0x0100 | (error_code as u16),
+            FpgaError::BitstreamError(error_code) => {
+                0x0200 | (error_code as u16)
+            }
+            FpgaError::InvalidState => 0x0300,
+            FpgaError::InvalidValue => 0x0301,
+            FpgaError::PortDisabled => 0x0400,
+        }
+    }
+}
+
+impl From<SpiError> for FpgaError {
+    fn from(e: SpiError) -> Self {
+        FpgaError::ImplError(u32::from(e) as u8)
+    }
+}
+
+impl From<FpgaError> for u32 {
+    fn from(e: FpgaError) -> Self {
+        u16::from(e) as u32
+    }
+}
+
+impl core::convert::TryFrom<u16> for FpgaError {
+    type Error = ();
+
+    fn try_from(v: u16) -> Result<Self, Self::Error> {
+        match v & 0xff00 {
+            0x0100 => Ok(FpgaError::ImplError((v & 0x00ff) as u8)),
+            _ => Err(()),
+        }
+    }
+}
+
+impl core::convert::TryFrom<u32> for FpgaError {
+    type Error = ();
+
+    fn try_from(v: u32) -> Result<Self, Self::Error> {
+        Self::try_from(v as u16)
+    }
+}
+
+#[derive(Copy, Clone, Debug, FromPrimitive, PartialEq, AsBytes)]
+#[repr(u8)]
+pub enum DeviceState {
+    Unknown = 0,
+    Disabled = 1,
+    AwaitingBitstream = 2,
+    RunningApplication = 3,
+    Error = 4,
+}
+
+#[derive(Copy, Clone, Debug, FromPrimitive, PartialEq, AsBytes)]
+#[repr(u8)]
+pub enum BitstreamType {
+    Uncompressed = 0,
+    Compressed = 1,
+}
+
+#[derive(Copy, Clone, Debug, FromPrimitive, PartialEq, AsBytes)]
+#[repr(u8)]
+pub enum WriteOp {
+    Write = 0,
```

Comment:
Nit: Why is there a gap here?

Reply:
Because the RTL implementation uses a more generic command type which has `Read` with value 1. But because the API is already split in read/write I wanted an Op type which wouldn't accidentally let you do the wrong thing.